### PR TITLE
Agent cache: detect DNS names and TTL changes in registration entries

### LINKED ISSUE #1425: Agent should detect updates to its cache and push them out
Changes to registration entries should be detected by the agent and the cache should be updated so the changes are reflected without having to wait a full rotation period.

### PR DESCRIPTION
**Pull request by amartinezfayo**
This PR adds the ability to push a new SVID as soon as a change is detected in the DNS names or TTL of a registration entry.
Fixes #1425.

### CODE REVIEW

File: pkg/agent/manager/cache/cache_test.go [hunk 1]
```diff
@@ -436,20 +437,24 @@ func TestCheckSVIDCallback(t *testing.T) {
 	// no calls because there are no registration entries
 	cache.Update(&Update{
 		Bundles: makeBundles(bundleV2),
-	}, func(entry *common.RegistrationEntry, svid *X509SVID) {
+	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) {
 		assert.Fail(t, "should not be called if there are no registration entries")
 	})
 
-	foo := makeRegistrationEntry("FOO")
+	foo := makeRegistrationEntry("FOO", 60)
```
Comment (azdagron) [inline pkg/agent/manager/cache/cache_test.go]: nit: it seems preferable to create another method (i.e. `makeRegistrationEntryWithTTL`) or something that allowed you to use a different TTL at this one spot than changing the existing function and passing the default `ttl` everywhere else...

File: pkg/agent/manager/cache/cache_test.go [hunk 2]
```diff
@@ -22,13 +22,14 @@ var (
 	bundleV3      = bundleutil.BundleFromRootCA("spiffe://domain.test", &x509.Certificate{Raw: []byte{3}})
 	otherBundleV1 = bundleutil.BundleFromRootCA("spiffe://otherdomain.test", &x509.Certificate{Raw: []byte{4}})
 	otherBundleV2 = bundleutil.BundleFromRootCA("spiffe://otherdomain.test", &x509.Certificate{Raw: []byte{5}})
+	ttl           = int32(600)
```
Comment (azdagron) [inline pkg/agent/manager/cache/cache_test.go]: nit: maybe `defaultTTL`?

File: pkg/agent/manager/sync.go [hunk 3]
```diff
@@ -167,3 +179,18 @@ func parseBundles(bundles map[string]*common.Bundle) (map[string]*cache.Bundle,
 	}
 	return out, nil
 }
+
+// equal determines whether two string slices are equal or not
+func equal(x, y []string) bool {
```
Comment (azdagron) [inline pkg/agent/manager/sync.go]: nit: `stringsEqual`?

**Comment by faisal-memon** [comment]
Verified that this is working with the latest build.

**Comment by amartinezfayo** [comment]
> Verified that this is working with the latest build.

Thanks @faisal-memon for verifying it!

**Comment by faisal-memon** [comment]
I did some further testing today and noticed that sometimes the update doesn't go through. I can see that the Spire Server has the entry correct but the SVID doesn't get pushed to the workload. Only suspicous thing I see in the logs is this from the agent:

```
time="2020-03-19T01:00:37Z" level=info msg="Renewing X509-SVID" spiffe_id="spiffe://example.org/ns/default/sa/bookinfo-reviews" subsystem_name=manager
time="2020-03-19T01:00:37Z" level=info msg="Renewing X509-SVID" spiffe_id="spiffe://example.org/ns/default/sa/bookinfo-reviews" subsystem_name=manager
time="2020-03-19T01:00:37Z" level=info msg="Renewing X509-SVID" spiffe_id="spiffe://example.org/ns/default/sa/bookinfo-reviews" subsystem_name=manager
time="2020-03-19T01:00:37Z" level=debug msg="PID attested to have selectors" pid=2412589 selectors="[type:\"unix\" value:\"uid:2102\"  type:\"unix\" value:\"gid:1000\"  type:\"k8s\" value:\"sa:bookinfo-reviews\"  type:\"k8s\" value:\"ns:default\"  type:\"k8s\" value:\"node-name:gke-fmemon-cluster-default-pool-0729ba70-fqzc\"  type:\"k8s\" value:\"pod-uid:0709f986-697d-11ea-a5bf-42010a8a0187\"  type:\"k8s\" value:\"pod-name:reviews-v1-5c5b7b9f8d-cm72f\"  type:\"k8s\" value:\"container-name:proxy\"  type:\"k8s\" value:\"container-image:gcr.io/nginx-mesh-dev/proxy-agent:fmemon\"  type:\"k8s\" value:\"pod-label:app:reviews\"  type:\"k8s\" value:\"pod-label:nginx.org/deployment:reviews-v1\"  type:\"k8s\" value:\"pod-label:pod-template-hash:5c5b7b9f8d\"  type:\"k8s\" value:\"pod-label:spiffe.io/spiffeid:reviews-v1-5c5b7b9f8d-cm72f-846c9649cb\"  type:\"k8s\" value:\"pod-label:spire-workload:reviews-v1\"  type:\"k8s\" value:\"pod-label:version:v1\"  type:\"k8s\" value:\"pod-owner:ReplicaSet:reviews-v1-5c5b7b9f8d\"  type:\"k8s\" value:\"pod-owner-uid:ReplicaSet:0704442d-697d-11ea-a5bf-42010a8a0187\" ]" subsystem_name=workload_api
time="2020-03-19T01:00:37Z" level=error msg="No identity issued" error="<nil>" registered=false subsystem_name=workload_api
```

**Comment by amartinezfayo** [comment]
So you see this working fine sometimes and sometimes it doesn't work?
When a change in a DNS name is detected, you should see an
`Updating SVIDs with outdated attributes in cache` debug message 
right before a `Renewing X509-SVID` message in the log. That should happen within 5 seconds of the entry change (default sync time interval).

Providing Agent and Server logs for a larger period of time could help to try to figure out what's happening.

**Comment by faisal-memon** [comment]
Yes, it works most of the time but sometimes a pod doesnt get the updated SVID. I noticed more often when im testing the bookinfo app which creates 6 svids and then updates them right away with the dns names. Ill do some more digging today and look out for those log messages.

**Comment by evan2645** [comment]
From the output above, I see two things. First, the workload connection to the agent - that is a new connection, perhaps the workload has just launched? Second, the agent reporting it has no SVID for the workload (`registered=false`)... I didn't have a close look at the patch, but is it possible it is doing some sort of remove-then-add?

It would be good to move this into a new/dedicated issue so we can track it and make sure it gets fixed up before our next release

**Comment by evan2645** [comment]
Also, thank you again for thoroughly testing this @faisal-memon. Once we get a reproduction, we should be sure to add a regression test somewhere

**Comment by faisal-memon** [comment]
I was able to reproduce and capture the full logs. Its the `bookinfo-details` thats not getting the update. I can see the `Update` and `Renewing` messages in the logs for the other bookinfo service but its missing for `bookinfo-details`.

[spire_server.log](https://github.com/spiffe/spire/files/4361606/spire_server.log)
[spire-agent-hzslb.log](https://github.com/spiffe/spire/files/4361607/spire-agent-hzslb.log)

**Pull request merged**